### ISSUE TITLE
dotnet TFM and Portable Class Libraries appear incompatible

### ISSUE THREAD

**Issue by tillig** [NuGet/Home#1625]
I'm unsure if this is a bug, but it's making life difficult so I figured I'd report it. ([My original blog entry about it is here.](http://www.paraesthesia.com/archive/2015/10/20/gotcha-with-dnx-dependency-resolution-dotnet-pcl/))

I'm also not entirely sure if this is a `dnu` issue or a NuGet issue.

We recently released [Autofac 4.0.0-beta8-157](https://www.nuget.org/packages/Autofac/4.0.0-beta8-157) to NuGet to coincide with the DNX beta 8 release. As part of that update, we re-added the classic PCL target `.NETPortable,Version=v4.5,Profile=Profile259` (which is `portable-net45+dnxcore50+win+wpa81+wp80+MonoAndroid10+Xamarin.iOS10+MonoTouch10`) because older VS versions and some project types were having trouble finding a compatible version of Autofac 4.0.0 - they didn't rectify the `dotnet` target framework as a match.

In working on a beta 8 compatible version of [Autofac.Configuration](https://github.com/autofac/Autofac.Configuration), which targeted `dnx451`, `dotnet`, and `net45` in beta 7, I figured we could just update to start using Autofac 4.0.0-beta8-157.

Instead, I started getting a lot of build errors when targeting the `dotnet` framework moniker.

```
Building Autofac.Configuration for .NETPlatform,Version=v5.0
  Using Project dependency Autofac.Configuration 4.0.0-beta8-1
    Source: E:\dev\opensource\Autofac\Autofac.Configuration\src\Autofac.Configuration\project.json

  Using Package dependency Autofac 4.0.0-beta8-157
    Source: C:\Users\tillig\.dnx\packages\Autofac\4.0.0-beta8-157
    File: lib\dotnet\Autofac.dll

  Using Package dependency Microsoft.Framework.Configuration 1.0.0-beta8
    Source: C:\Users\tillig\.dnx\packages\Microsoft.Framework.Configuration\1.0.0-beta8
    File: lib\dotnet\Microsoft.Framework.Configuration.dll

  Using Package dependency Microsoft.Framework.Configuration.Abstractions 1.0.0-beta8
    Source: C:\Users\tillig\.dnx\packages\Microsoft.Framework.Configuration.Abstractions\1.0.0-beta8
    File: lib\dotnet\Microsoft.Framework.Configuration.Abstractions.dll

  Using Package dependency System.Collections 4.0.11-beta-23409
    Source: C:\Users\tillig\.dnx\packages\System.Collections\4.0.11-beta-23409
    File: ref\dotnet\System.Collections.dll

  (...and some more package dependencies that got resolved, then...)

  Unable to resolve dependency fx/System.Collections

  Unable to resolve dependency fx/System.ComponentModel

  Unable to resolve dependency fx/System.Core

  (...and a lot more fx/* items unresolvable.)
```

This was, at best, confusing. I mean, in the same target framework, I see these two things together:

```
  Using Package dependency System.Collections 4.0.11-beta-23409
    Source: C:\Users\tillig\.dnx\packages\System.Collections\4.0.11-beta-23409
    File: ref\dotnet\System.Collections.dll

  Unable to resolve dependency fx/System.Collections
```

So it _found_ System.Collections, but it _didn't find_ System.Collections? Running the `dnu list --details` command, I saw this:

```
[Target framework .NETPlatform,Version=v5.0 (dotnet)]

Framework references:
  fx/System.Collections  - Unresolved
    by Package: Autofac 4.0.0-beta8-157

  fx/System.ComponentModel  - Unresolved
    by Package: Autofac 4.0.0-beta8-157

  (...and a bunch more of these...)


Package references:
* Autofac 4.0.0-beta8-157
    by Project: Autofac.Configuration 4.0.0-beta8-1

* Microsoft.Framework.Configuration 1.0.0-beta8
    by Project: Autofac.Configuration 4.0.0-beta8-1

  Microsoft.Framework.Configuration.Abstractions 1.0.0-beta8
    by Package: Microsoft.Framework.Configuration 1.0.0-beta8

  System.Collections 4.0.11-beta-23409
    by Package: Autofac 4.0.0-beta8-157...
    by Project: Autofac.Configuration 4.0.0-beta8-1

  (...and so on.)
```

What I see is that Autofac 4.0.0-beta8-157 needs _both_ the framework assembly _and_ the dependency package for `System.Collections` which seems weird.

Looking in the generated `.nuspec` file for the updated core Autofac, I see:

``` xml
<?xml version="1.0"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <!-- ... -->
    <dependencies>
      <group targetFramework="DNX4.5.1" />
      <group targetFramework=".NETPlatform5.0">
        <dependency id="System.Collections" version="4.0.11-beta-23409" />
        <dependency id="System.Collections.Concurrent" version="4.0.11-beta-23409" />
        <!-- ... -->
      </group>
      <group targetFramework=".NETFramework4.5" />
      <group targetFramework=".NETCore4.5" />
      <group targetFramework=".NETPortable4.5-Profile259" />
    </dependencies>
    <frameworkAssemblies>
      <!-- ... -->
      <frameworkAssembly assemblyName="System.Collections" targetFramework=".NETPortable4.5-Profile259" />
      <frameworkAssembly assemblyName="System.ComponentModel" targetFramework=".NETPortable4.5-Profile259" />
      <frameworkAssembly assemblyName="System.Core" targetFramework=".NETPortable4.5-Profile259" />
      <frameworkAssembly assemblyName="System.Diagnostics.Contracts" targetFramework=".NETPortable4.5-Profile259" />
      <frameworkAssembly assemblyName="System.Diagnostics.Debug" targetFramework=".NETPortable4.5-Profile259" />
      <frameworkAssembly assemblyName="System.Diagnostics.Tools" targetFramework=".NETPortable4.5-Profile259" />
      <!-- ... -->
    </frameworkAssemblies>
  </metadata>
</package>
```

**The list of failed `fx/*` dependencies is exactly the same as the list of `frameworkAssembly` references that target `.NETPortable4.5-Profile259` in the `.nuspec`.**

By removing the `dotnet` target framework moniker from Autofac.Configuration and compiling for _specific targets_, everything resolves correctly.

Again, I'm not sure if this is the desired behavior, but it seems if you have a library that takes a dependency on something that targets _both_ `dotnet` and a PCL profile, you suddenly can't support `dotnet` unless you _also_ support the PCL profile.

If you'd like to try this and see it in action, check out [the Autofac.Configuration repo at `14c10b5bf6`](https://github.com/autofac/Autofac.Configuration/tree/14c10b5bf6216b1bdfb1f5414105161eeba06f9f) and run the `build.ps1` build script.

**If this is a `dnu` issue, let me know so I can post the issue over on the aspnet repo.**

**Comment by yishaigalatzer** [comment]
1. This should have been posted on the ASP.NET site, but we plan to start consolidated the two together, so we can keep it here.
2. This is by design, although a rather limiting design. So completely expected.
3. What are we doing about it, we are working on an alternative, for now removing dotnet is not a bad idea for your scenario.

@davidfowl

**Comment by tillig** [comment]
Thanks for getting back to me, @yishaigalatzer. Hopefully the alternative you and @davidfowl are working on will allow me to enable support for Universal Windows Platform. Right now, while core Autofac, Microsoft.Framework.Configuration, and the base framework dependencies I'm using support UWP/UAP, I can't seem to add a TFM to the Autofac.Configuration package that will allow it to also support UWP/UAP.

**Comment by yishaigalatzer** [comment]
You can just add uap10 without adding dotnet

Sent from my Windows Phone

**Comment by tillig** [comment]
uap10 fails because it doesn't think either Microsoft.Framework.Configuration or Autofac core support it, though they clearly do.

**Comment by tillig** [comment]
I also tried adding netcore50 and while I can get Autofac.Configuration to compile under that, I can't seem to add a reference to it from a UWP project because it's not seen as compatible.

**Comment by yishaigalatzer** [comment]
@ericstj @davidfowl. This should work just fine, we need to figure out why these are not compatible, and figure out what to change (the packages, nuget or something else

Sent from my Windows Phone

**Comment by tillig** [comment]
For reference, here's the error when compiling using `uap10`:

![Error when targeting uap10](https://cloud.githubusercontent.com/assets/1156571/10696337/02f411f4-795c-11e5-8727-91023f2de00e.PNG)

**Comment by ericstj** [comment]
I think the problem is your project's target moniker.  That should be UAP version 10.0 not 1.0.  What is in your project.json frameworks section?  Should be `UAP10.0`.

**Comment by tillig** [comment]
If I switch to `uap10.0` I do see that it's building for UAP 10:

```
Building Autofac.Configuration for UAP,Version=v10.0
```

But with a `project.json` section like this (non-dependencies trimmed):

``` json
{
    "dependencies": {
        "Autofac": "4.0.0-beta8-*",
        "Microsoft.Framework.Configuration": "1.0.0-beta8"
    },
    "frameworks": {
        "uap10.0": {
            "dependencies": {
                "System.Collections": "4.0.11-beta-*",
                "System.ComponentModel.TypeConverter": "4.0.1-beta-*",
                "System.Diagnostics.Tools": "4.0.1-beta-*",
                "System.Globalization": "4.0.11-beta-*",
                "System.IO": "4.0.11-beta-*",
                "System.Linq": "4.0.1-beta-*",
                "System.Reflection": "4.1.0-beta-*",
                "System.Reflection.Extensions": "4.0.1-beta-*",
                "System.Reflection.TypeExtensions": "4.0.1-beta-*",
                "System.Resources.ResourceManager": "4.0.1-beta-*"
            }
        }
    }
}
```

**I end up with the same set of errors that I had when I target `dotnet`** - a series of things where it finds the package dependencies but can't resolve the `fx/*` counterparts:

```
  Using Package dependency System.Collections 4.0.11-beta-23409
    Source: C:\Users\tillig\.dnx\packages\System.Collections\4.0.11-beta-23409
    File: ref\dotnet\System.Collections.dll

...

  Unable to resolve dependency fx/System.Collections
```

**Comment by ericstj** [comment]
Let's take a step back.  DNX cannot build UAP apps so this configuration is not really valid.  There are a lot of things wrong with that project.json  (missing runtime section, missing runtime packages) for UAP but let's not go there.

Can you restate what it is you are trying to do and show me your original project.json?

**Comment by tillig** [comment]
The original statement remains the same as in the very first post, but for clarity:
- Autofac is a DNX project that builds targeting `dnx451`, `dotnet`, `net45`, `netcore45`, and `.NETPortable,Version=v4.5,Profile=Profile259`. We found that to maintain compatibility with all the various project types we were previously compatible with _and_ to enable earlier VS versions to work, we needed to target all of these things.
- Microsoft.Framework.Configuration is a DNX project that targets `net45`, `dnx451`, `dnxcore50`, and `dotnet`.
- Both Autofac and Microsoft.Framework.Configuration are compatible with UWP.
- All of the dependencies of both Autofac and Microsoft.Framework.Configuration are compatible with UWP.
- Autofac.Configuration is a DNX project that takes a reference on Autofac and Microsoft.Framework.Configuration.
- **I want to make Autofac.Configuration compatible with UWP.**

My challenge is that I can't figure out the right incantation to chant over DNX to allow this to happen.

As originally posted, the whole project with the reproduction is [in the Autofac.Configuration repo at `14c10b5bf6`](https://github.com/autofac/Autofac.Configuration/tree/14c10b5bf6216b1bdfb1f5414105161eeba06f9f).

For clarity/reference, the full `project.json` that I can't get to work - from that commit - is:

``` json
{
    "version": "4.0.0-beta8-*",
    "authors": ["Autofac Contributors"],
    "description": "Configuration support for Autofac",
    "summary": "Enables Autofac dependencies to be set up using configuration sources.",
    "projectUrl": "http://autofac.org",
    "licenseUrl": "http://www.opensource.org/licenses/mit-license.php",
    "iconUrl": "http://code.google.com/p/autofac/logo",
    "requireLicenseAcceptance": false,
    "tags": [ "Autofac", "DI", "IoC" ],
    "compilationOptions": {
        "warningsAsErrors": true
    },
    "compile": ["**/*.cs"],
    "resource": ["**/*.resx"],
    "repository": {
        "type": "git",
        "url": "https://github.com/autofac/Autofac.Configuration"
    },
    "dependencies": {
        "Autofac": "4.0.0-beta8-*",
        "Microsoft.Framework.Configuration": "1.0.0-beta8"
    },
    "frameworks": {
        "dnx451": {},
        "dnxcore50": {
            "dependencies": {
                "System.Collections": "4.0.11-beta-*",
                "System.ComponentModel.TypeConverter": "4.0.1-beta-*",
                "System.Diagnostics.Tools": "4.0.1-beta-*",
                "System.Globalization": "4.0.11-beta-*",
                "System.IO": "4.0.11-beta-*",
                "System.Linq": "4.0.1-beta-*",
                "System.Reflection": "4.1.0-beta-*",
                "System.Reflection.Extensions": "4.0.1-beta-*",
                "System.Reflection.TypeExtensions": "4.0.1-beta-*",
                "System.Resources.ResourceManager": "4.0.1-beta-*"
            }
        },
        "dotnet": {
            "dependencies": {
                "System.Collections": "4.0.11-beta-*",
                "System.ComponentModel.TypeConverter": "4.0.1-beta-*",
                "System.Diagnostics.Tools": "4.0.1-beta-*",
                "System.Globalization": "4.0.11-beta-*",
                "System.IO": "4.0.11-beta-*",
                "System.Linq": "4.0.1-beta-*",
                "System.Reflection": "4.1.0-beta-*",
                "System.Reflection.Extensions": "4.0.1-beta-*",
                "System.Reflection.TypeExtensions": "4.0.1-beta-*",
                "System.Resources.ResourceManager": "4.0.1-beta-*"
            }
        },
        "net45":{}
    }
}
```

I admit **I could very well be doing something wrong**, but there is... _less than stellar/complete documentation_ on the meanings and permutations of all of these things, so there's a lot of "stabbing in the dark," if you will. Any guidance/assistance is greatly appreciated.

**Comment by davkean** [comment]
Yep, this is all badly documented.

The set of frameworks you are targeting is somewhat confusing; dnx451, dotnet, net45, netcore45, and .NETPortable,Version=v4.5,Profile=Profile259.

How did you choose these ones?

The portable profile (259) that you are targeting supports .NET Framework 4.5+, ASP.NET Core 5.0, Windows 8+, Windows Phone 8.1+ and Windows Phone Silverlight 8+. This already supports UWP ("Windows 10" & "DNXCore").

Are you using WinRT at all? If not, then you don't need netcore45, netcore45 is Windows 8+, which you are already targeting in portable.
Are you using a dnx-only feature when targeting dnx451? Then you probably don't need dnx451.

It sounds like you really have two unique builds of this assembly, one that builds again Microsoft.Framework.Configuration and one that doesn't? Would that be right?

**Comment by tillig** [comment]
From a level-setting standpoint: Autofac is an inversion-of-control container for .NET. It's like a tool or framework library. It's not, itself, an _application_.

The reason I say that is that I think there may be a misunderstanding or a slight miscommunication, which could be just me. Autofac _is used by people on these platforms_ but Autofac proper _does not, itself_, "use a platform." So when you ask if "I'm using WinRT," the answer is, "No, **I'm** not using WinRT, but the **people that consume Autofac** are." This same answer goes for the other platforms.

We chose the **Autofac core** (ie `Autofac.dll`) frameworks by looking at profile 259 - the profile we supported before NuGet 3 and the new TFMs - and trying to approximate it with the new TFMs. The problem we ran into was that if we _only_ used `dnx451`, `dotnet`, `net45`, `netcore45` that VS 2013 and some other project types would fail to find a compatible Autofac despite actually being compatible. By adding the profile 259 back in, we got compatibility with the older VS and project types that weren't previously finding them. For example, profile 259 allowed us to be compatible with Xamarin.

When _only_ targeting the portable profile, we had challenges with newer project types finding a compatible Autofac. This combination of profiles seemed to be the magic combination that allowed us to continue supporting the platforms we had supported before the new TFMs were introduced.

I've tried several combinations of the new TFMs in conjunction with the portable profile 259 for core Autofac and if I remove any of the new TFMs, I've generally found a project type that stops finding a compatible Autofac. I don't have comprehensive doc on which project types or target frameworks stop working for each of the TFMs. I should probably have kept better records on that, but I didn't and it's too late now. If it's not _hurting_ anything, I'd just as soon leave it; but if you think the extra targets are contributing to the problem I'm seeing, I can definitely go back and start documenting more specifically what I'm seeing.

**Building core Autofac (`Autofac.dll`) works great.**. Core Autofac has its own build, its own repo, etc. It's a separate thing. The problem is in building Autofac.Configuration.

**The Autofac.Configuration project is an extension on Autofac.** Autofac.Configuration (ie, `Autofac.Configuration.dll`) is also a tool/framework library. It allows you to use a configuration file to specify dependencies that get registered into your Autofac IoC container.

There is _one build_ of Autofac.Configuration. It takes a dependency on core Autofac and on Microsoft.Framework.Configuration.

**Autofac.Configuration does not target PCL profile 259.** Core Autofac does. Autofac.Configuration does not target profile 259 because Microsoft.Framework.Configuration is not seen as compatible with profile 259.

My goal is to allow Autofac.Configuration to be compatible with the subset of frameworks that both core Autofac and Microsoft.Framework.Configuration support. I would like to achieve maximum compatibility.

This is where I'm failing: I can't get Autofac.Configuration to support UWP despite all of its dependencies supporting that framework. What that means is that I can't offer configuration support to people using the Autofac IoC container on that platform, and I'd really, really like to. I would love to know how to make that happen.
